src/amqp.c

```diff
@@ -208,11 +216,26 @@ static int start_connection(struct settings *settings)
 	if (r.reply_type != AMQP_RESPONSE_NORMAL) {
 		amqp_connection_close(amqp_ctx.conn, AMQP_REPLY_SUCCESS);
 		amqp_destroy_connection(amqp_ctx.conn);
+		amqp_ctx.conn = NULL;
 		hal_log_error("amqp_channel_open(): %s",
 				amqp_rpc_reply_string(r));
 		return -1;
 	}
 
+	amqp_ctx.amqp_io = l_io_new(amqp_get_sockfd(amqp_ctx.conn));
+
+	status = l_io_set_disconnect_handler(amqp_ctx.amqp_io, on_disconnect,
+					  NULL, NULL);
+	if (!status) {
+		amqp_connection_close(amqp_ctx.conn, AMQP_REPLY_SUCCESS);
+		amqp_destroy_connection(amqp_ctx.conn);
+		l_io_destroy(amqp_ctx.amqp_io);
+		amqp_ctx.conn = NULL;
+		amqp_ctx.amqp_io = NULL;
+		hal_log_error("Error on set up disconnect handler");
+		return -1;
+	}
+
```

Comment:
What do you think about using 'goto' for the not expect return (-1) on this function?

Reply:
I didn't understand. Can you explain? On the previous commit (b1d2d5435e7133305f5f612180fb668205af676b) the function start_timeout became a timeout callback. So we no longer can return a value.